kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeature.java

```diff
@@ -0,0 +1,226 @@
+ * table that uses the feature).
+ */
+public abstract class TableFeature {
+
+  /////////////////////////////////////////////////////////////////////////////////
+  /// Instance variables.                                                       ///
+  /////////////////////////////////////////////////////////////////////////////////
+  private final String featureName;
+  private final int minReaderVersion;
+  private final int minWriterVersion;
+
+  /////////////////////////////////////////////////////////////////////////////////
+  /// Public methods.                                                           ///
+  /////////////////////////////////////////////////////////////////////////////////
+  /**
+   * Constructor. Does validations to make sure:
+   *
+   * <ul>
+   *   <li>Feature name is not null or empty and has valid characters
+   *   <li>minReaderVersion is always 0 for writer features
+   *   <li>all legacy features can be auto applied based on the metadata and protocol
+   *
+   * @param featureName a globally-unique string indicator to represent the feature. All characters
+   *     must be letters (a-z, A-Z), digits (0-9), '-', or '_'. Words must be in camelCase.
+   * @param minReaderVersion the minimum reader version this feature requires. For a feature that
+   *     can only be explicitly supported, this is either `0` (i.e writerOnly feature) or `3` (the
+   *     reader protocol version that supports table features), depending on the feature is
+   *     writer-only or reader-writer. For a legacy feature that can be implicitly supported, this
+   *     is the first protocol version which the feature is introduced.
+   * @param minWriterVersion the minimum writer version this feature requires. For a feature that
+   *     can only be explicitly supported, this is the writer protocol `7` that supports table
+   *     features. For a legacy feature that can be implicitly supported, this is the first protocol
+   *     version which the feature is introduced.
+   */
+  public TableFeature(String featureName, int minReaderVersion, int minWriterVersion) {
+    this.featureName = requireNonNull(featureName, "name is null");
+    checkArgument(!featureName.isEmpty(), "name is empty");
+    checkArgument(
+        featureName.chars().allMatch(c -> Character.isLetterOrDigit(c) || c == '-' || c == '_'),
+        "name contains invalid characters: " + featureName);
+    checkArgument(minReaderVersion >= 0, "minReaderVersion is negative");
+    checkArgument(minWriterVersion >= 1, "minWriterVersion is less than 1");
+    this.minReaderVersion = minReaderVersion;
+    this.minWriterVersion = minWriterVersion;
+
+    validate();
+  }
+
+  /** @return the name of the table feature. */
+  public String featureName() {
+    return featureName;
+  }
+
+  /**
+   * @return true if this feature is applicable to both reader and writer, false if it is
+   *     writer-only.
+   */
+  public boolean isReaderWriterFeature() {
```

Comment:
Q: I see, so there's no `isWriterFeature` since _every_ feature is a writer feature, right?

Reply:
yep, thats correct.